src/alpaka/AlpakaCore/device_unique_ptr.h

```diff
@@ -0,0 +1,72 @@
+#ifndef HeterogeneousCore_AlpakaUtilities_interface_device_unique_ptr_h
+#define HeterogeneousCore_AlpakaUtilities_interface_device_unique_ptr_h
+
+#include <memory>
+#include <type_traits>
+
+#include "AlpakaCore/host_unique_ptr.h"
+
+namespace cms {
+  namespace alpakatools {
+    namespace device {
+      namespace impl {
+        template <typename TData>
+        class DeviceDeleter {
+        public:
+          DeviceDeleter(::ALPAKA_ACCELERATOR_NAMESPACE::AlpakaDeviceBuf<TData> buffer) : buf{std::move(buffer)} {}
+
+          void operator()(void* d_ptr) {
+            if constexpr (allocator::policy == allocator::Policy::Caching) {
+              if (d_ptr) {
+                allocator::getCachingDeviceAllocator().DeviceFree(buf);
+              }
+            }
+          }
+
+        private:
+          ::ALPAKA_ACCELERATOR_NAMESPACE::AlpakaDeviceBuf<TData> buf;
+        };
+      }  // namespace impl
+      template <typename TData>
+      using unique_ptr =
+#ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
+          std::unique_ptr<TData,
+                          impl::DeviceDeleter<
+                              std::conditional_t<allocator::policy == allocator::Policy::Caching, std::byte, TData>>>;
+#else
+          host::unique_ptr<TData>;
+#endif
+    }  // namespace device
```

Comment:
Technically this violates ODR. Templating the `unique_ptr` also over Device and providing a specialization for `DevCudaRt` if `ALPAKA_ACC_GPU_CUDA_ENABLED` is defined would work. In longer term, I think, we should think about hiding the exact deleter type from the type of the `unique_ptr` (or just go directly to `shared_ptr` semantics).